File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/manager/BaseAssetManager.java
package com.greentree.engine.moon.assets.serializator.manager;

import com.greentree.commons.reflection.info.TypeInfo;
import com.greentree.engine.moon.assets.asset.Asset;
import com.greentree.engine.moon.assets.asset.ThrowAsset;
import com.greentree.engine.moon.assets.key.AssetKey;
import com.greentree.engine.moon.assets.serializator.AssetSerializator;
import com.greentree.engine.moon.assets.serializator.manager.cache.WeakHashMapCache;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.message.MapMessage;

import java.util.HashMap;
import java.util.function.Function;

public final class BaseAssetManager implements MutableAssetManager {

    private static final Logger LOG = LogManager.getLogger(BaseAssetManager.class);

    //    private final AssetSerializatorContainer container = new AssetSerializatorContainer(HashMapCache::new);
    private final AssetSerializatorContainer container = new AssetSerializatorContainer(WeakHashMapCache::new);

    public <T> Asset<T> load(TypeInfo<T> type, AssetKey key) {
        Asset<T> result;
        try {
            result = container.load(this, type, key);
        } catch (RuntimeException e) {
//            LOG.info(ASSETS, "asset loader throw. info: " + info, e);
//            return NotValidAsset.instance();
            return new ThrowAsset(e);
        }
        LOG.debug(ASSETS, () -> new MapMessage<>(new HashMap<>() {{
            put("result", result);
            put("type", type);
            put("key", key);
        }}));
        return result;
    }
    
    @Override
    public <T> void addSerializator(AssetSerializator<T> serializator) {
        container.addSerializator(serializator);
    }

    @Override
    public void addGenerator(
            Function<? super TypeInfo<?>, ? extends AssetSerializator<?>> generator) {
        container.addGenerator(generator);
    }

}

File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/manager/BaseAssetManager.kt
package com.greentree.engine.moon.assets.serializator.manager

import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.engine.moon.assets.asset.Asset
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.serializator.loader.AssetLoader
import com.greentree.engine.moon.assets.serializator.loader.CacheAssetLoader
import com.greentree.engine.moon.assets.serializator.loader.DefaultAssetLoader
import com.greentree.engine.moon.assets.serializator.loader.MultiAssetLoader
import com.greentree.engine.moon.assets.serializator.loader.ResultAssetLoader
import com.greentree.engine.moon.assets.serializator.manager.cache.WeakHashMapCacheFactory
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.message.SimpleMessage

class BaseAssetManager : MutableAssetManager {

	private val loaders = mutableListOf<AssetLoader>()
	private val multiLoaders =
		CacheAssetLoader(
			MultiAssetLoader(loaders),
			WeakHashMapCacheFactory()
		)

	init {
		addLoader(DefaultAssetLoader())
		addLoader(ResultAssetLoader())
	}

	private inner class AssetLoaderContext : AssetLoader.Context {

		override fun <T : Any> load(type: TypeInfo<T>, key: AssetKey) = this@BaseAssetManager.load(type, key)
	}

	private val context = AssetLoaderContext()

	override fun <T : Any> load(type: TypeInfo<T>, key: AssetKey): Asset<T> {
		val result = multiLoaders.load(context, type, key)
		LOG.debug(AssetManager.ASSETS) { SimpleMessage("type: $type key: $key result: $result") }
		return result
	}

	override fun addLoader(loader: AssetLoader) {
		loaders.add(loader)
	}

	companion object {

		private val LOG = LogManager.getLogger(
			BaseAssetManager::class.java
		)
	}
}